From 3b70b6ffcbe3259671d07fff9f7bc7bbbe833dae Mon Sep 17 00:00:00 2001
From: Marek Vasut <marex@denx.de>
Date: Tue, 26 Jul 2022 01:21:18 +0200
Subject: [PATCH 17/60] drm/bridge: tc358767: Attach bridge in bridge attach
 callback again

The NXP downstream DSIM driver cannot deal with modern attachment of DSI
bridges in their probe callback and still expects bridge attachment to
happen in bridge attach callback like it used to be in the past. Undo the
bridge attachment in probe and move it back into bridge attach callback
to make this driver compatible with the downstream DSIM driver.

Upstream-Status: Inappropriate [Workaround for deprecated NXP DSIM driver]
Signed-off-by: Marek Vasut <marex@denx.de>
---
 drivers/gpu/drm/bridge/tc358767.c | 15 +++++----------
 1 file changed, 5 insertions(+), 10 deletions(-)

diff --git a/drivers/gpu/drm/bridge/tc358767.c b/drivers/gpu/drm/bridge/tc358767.c
index 256db9ec5a6de..ac905f0c655cc 100644
--- a/drivers/gpu/drm/bridge/tc358767.c
+++ b/drivers/gpu/drm/bridge/tc358767.c
@@ -1666,13 +1666,16 @@ static const struct drm_connector_funcs tc_connector_funcs = {
 	.atomic_destroy_state = drm_atomic_helper_connector_destroy_state,
 };
 
+static int tc_mipi_dsi_host_attach(struct tc_data *tc);
 static int tc_dpi_bridge_attach(struct drm_bridge *bridge,
 				enum drm_bridge_attach_flags flags)
 {
 	struct tc_data *tc = bridge_to_tc(bridge);
+	int ret;
 
-	if (!tc->panel_bridge)
-		return 0;
+	ret = tc_mipi_dsi_host_attach(tc);
+	if (ret)
+		return ret;
 
 	return drm_bridge_attach(tc->bridge.encoder, tc->panel_bridge,
 				 &tc->bridge, flags);
@@ -2184,14 +2187,6 @@ static int tc_probe(struct i2c_client *client, const struct i2c_device_id *id)
 
 	i2c_set_clientdata(client, tc);
 
-	if (tc->input_connector_dsi) {			/* DSI input */
-		ret = tc_mipi_dsi_host_attach(tc);
-		if (ret) {
-			drm_bridge_remove(&tc->bridge);
-			return ret;
-		}
-	}
-
 	return 0;
 }
 
-- 
2.35.1

